### PR TITLE
Decorate default Symfony session listener for Symfony 3.4+

### PR DESCRIPTION
**Pull request by Toflar**
Fix for https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/issues/433.

### CODE REVIEW

File: src/EventListener/SessionListener.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\EventListener\SessionListener as BaseSessionListener;
+
+/**
+ * Decorates the default Symfony session listener.
+ *
+ * The default Symfony session listener automatically makes responses private
+ * in case the session was started. This kills the user context feature of
+ * FOSCache. We disable the default behaviour only if the user context header
```
Comment (dbu) [inline src/EventListener/SessionListener.php]: s/FOSCache/FOSHttpCache

File: src/EventListener/SessionListener.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * @var array
+     */
+    private $userIdentifierHeaders;
+
+    public function __construct(BaseSessionListener $inner, string $userHashHeader, array $userIdentifierHeaders)
+    {
+        $this->inner = $inner;
+        $this->userHashHeader = strtolower($userHashHeader);
+        $this->userIdentifierHeaders = array_map('strtolower', $userIdentifierHeaders);
```
Comment (dbu) [inline src/EventListener/SessionListener.php]: given that this is executed on every single request, i think we should do this in the DI extension class instead.

**Comment by Toflar** [comment]
Comments addressed, tests fixed :)

**Comment by dbu** [comment]
@andrerom could you maybe have a look on this? and do you have a setup where you can test if it actually works in a real environment?

**Comment by dbu** [comment]
we manually verified that the listener does what we expect from it.

we will proceed, worst case there will be a 2.2.1 bugfix release...

**Comment by emodric** [comment]
@dbu How will this handle Symfony 4.1 where the inner listener also does some other stuff in addition to setting caching headers: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php#L70-L97 ?

**Comment by dbu** [comment]
argh. we went back and forth between calling session->save ourselves and then call the parent, or doing what we do now. we decided for this in the end because things would otherwise break if the SaveSessionListener would start doing additional things.

as they found it acceptable to call `save` in the session listener, we should thus do the same. and have to, as currently the session is not saved anymore with sf 4.1

**Comment by dbu** [comment]
i created #437 and once somebody did that, will release 2.2.1

thanks a lot for having verified that @emodric !

**Comment by dbu** [comment]
@emodric can you check #438 please?

**Pull request merged**